docs/validator.md

```diff
@@ -0,0 +1,205 @@
+dereferenced.
+
+# Components
+
+## gpii.schema.validator.ajv
+
+The common grade, designed for use on the client and server-side.
+
+### Component Options
+
+The following component configuration options are supported:
+
+| Option             | Type     | Description |
+| ------------------ | -------- | ----------- |
+| `validatorOptions` | `Object` | Options to be passed to our AJV validator instance (check [the AJV options documentation](https://github.com/epoberezkin/ajv#options) for the syntax and available options). |
+
+
+### Invokers
+
+#### `{validator}.validate(key, content)`
+
+* `key {String}`:  The schema key we are validating against.
+* `content {Object}`: The JSON content we are validating.
+* Returns: An `{Object}` describing the errors, or `undefined` if there are no errors.
+
+The raw output provided by AJV is run through `sanitizeValidationErrors` (see below) to provide more user-friendly
+feedback where possible.
+
+### Static Functions
+
+
+#### `gpii.schema.validator.ajv.sanitizeValidationErrors(schemaKey, errors)`
+
+* `schemaKey {String}`: The schema key we are validating against.
+* `errors`: A map of existing error results to be sanitized (see below for examples).
+
+AJV gives us output like:
+
+```
+ [
+     {
+         "keyword": "minLength",
+         "schemaPath": "/password/minLength",
+         "dataPath": ".password",
+         "message": "should NOT be shorter than 8 characters"
+     },
+     {
+         "keyword": "pattern",
+         "schemaPath": "/password/allOf/1/pattern",
+         "dataPath": ".password",
+         "message": "should match pattern \"[A-Z]+\""
+     },
+     {
+         "keyword": "required",
+         "schemaPath": "/deep/required/0",
+         "dataPath": ".deep.required",
+         "message": "is a required property"
+     }
+ ]
+```
+
+This function uses [`gpii.schema.errors.evolveError`](evolveErrors.md) to look for `errors` definitions that match the `schemaPath` output
+from the validator. If it finds them, it replaces `message` with alternate wording, so that we end up with output like:
+
+```
+ [
+     {
+         "keyword": "minLength",
+         "schemaPath": "/password/minLength",
+         "dataPath": ".password",
+         "message": "Passwords must be at least 8 characters long."
+     },
+     {
+         "keyword": "pattern",
+         "schemaPath": "/password/allOf/1/pattern",
+         "dataPath": ".password",
+         "message": "Passwords must contain at least one uppercase character."
+     },
+     {
+         "keyword": "required",
+         "schemaPath": "/deep/required/0",
+         "dataPath": ".deep.required",
+         "message": "You must enter a value for the 'deeply required' field."
+     }
+ ]
+```
+See [the documentation for `evolveError`](evolveErrors.md) for more details.
+
+Note that `dataPath` values are ["Javascript property access notation"](https://github.com/epoberezkin/ajv#validation-errors).
```

Comment:
I have updated the docs as suggested.  I also created a ticket for [the follow-up work to better support working with both notations](https://issues.gpii.net/browse/GPII-2152).  IMO this would be a good one to work on the next time the project is "under the knife" (such as when we get around to [future-proofing the internationalisation of error messages](https://issues.gpii.net/browse/GPII-2113)).